File: src/shapes/triangle.rs
use image::{ImageBuffer, Rgba, RgbaImage, load_from_memory};
use std::marker::PhantomData;
use raqote::{Source, SolidSource, GradientStop, Point, Gradient, Spread, PathBuilder, DrawTarget, LineCap, LineJoin, DrawOptions, StrokeStyle};
use super::super::{DrawType, GradientSelect, Shape, Color};

pub fn draw_triangle_mut(img: &mut ImageBuffer<Rgba<u8>, Vec<u8>>, shape: &Shape) {
    let (x1, y1, x2, y2, x3, y3) = (shape.data[0], shape.data[1], shape.data[2], shape.data[3], shape.data[4], shape.data[5]);

    let x: f32;
    let y: f32;
    if x1 > x2 && x1 > x3{
        x = x1;
    }
    else if x2 > x1 && x2 > x3{
        x = x2;
    }
    else {
        x = x3;
    }
    if y1 > y2 && y1 > y3{
        y = y1;
    }
    else if y2 > y1 && y2 > y3{
        y = y2;
    }
    else {
        y = y3;
    }
    let source: Source;
    match &shape.color {
        Color::RGBA { red, green, blue, alpha } => {
            source = Source::Solid(SolidSource{r: *red, g: *green, b: *blue, a: *alpha});
        }
        Color::Gradients { gradients } => {
            let mut stops:Vec<GradientStop> = Vec::new();
            for gradient in gradients.iter() {
                stops.push(GradientStop {position: gradient.position, color: raqote::Color::new(gradient.color[3], gradient.color[0], gradient.color[1], gradient.color[2])});
            }
            match &gradients[0].gradient_select {
                GradientSelect::Linear {start, end} => {
                    let start = Point { x: start.x, y: start.y, _unit: PhantomData };
                    let end = Point { x: end.x, y: end.y, _unit: PhantomData };
                    source = Source::new_linear_gradient(Gradient { stops }, start, end, Spread::Pad );
                }
                GradientSelect::Radial{center, radius} => {
                    let start = Point { x: center.x, y: center.y, _unit: PhantomData };
                    source = Source::new_radial_gradient(Gradient { stops }, start, *radius, Spread::Pad );
                }
            }
        }
    }
    let mut path_builder = PathBuilder::new();
    path_builder.move_to(x1, y1);
    path_builder.line_to(x2, y2);
    path_builder.line_to(x3, y3);
    path_builder.close();
    let mut draw_target = DrawTarget::new(x as i32,y as i32);
    match shape.draw_type {
        DrawType::Fill => {
            draw_target.fill(&path_builder.finish(), &source, &DrawOptions::new());
        }
        DrawType::Stroke => {
            let style = StrokeStyle {
                width: 1.,
                cap: LineCap::Butt,
                join: LineJoin::Miter,
                miter_limit: 10.,
                dash_array: Vec::new(),
                dash_offset: 0.,
            };
            draw_target.stroke(&path_builder.finish(), &source, &style, &DrawOptions::new());
        }
    }
    let raw_image = draw_target.get_data_u8().to_vec();
    let buffer = load_from_memory(&raw_image).unwrap().to_rgba8();
    for (_x, _y, pixel,) in buffer.enumerate_pixels() {
        if pixel != &Rgba([0,0,0,0]) {
            img.put_pixel(_x, _y, pixel.to_owned());
        }
    }
}

File: lib/shapes/triangle.ts
import { Shape } from "../ffi.ts";
import { TriangleOptions, TriangleReturn } from "../types.ts";
import { setAngle, setType, setRGB, setRGBA, setColor, setGradient } from "../utilty.ts";

export default function(data: Shape[], options?: TriangleOptions): TriangleReturn {
    if(!options) options ={};
    if(!options.x1) options.x1 = 10;
    if(!options.y1) options.y1 = 1;
    if(!options.x2) options.x2 = 20;
    if(!options.y2) options.y2 = 5;
    if(!options.x3) options.x3 = 30;
    if(!options.y3) options.y3 = 10;
    if(!options.color) options.color = {red: 0, green: 0, blue:0, alpha: 255};
    if(!options.type) options.type = "fill";
    return { 
        x1:options.x1,
        y1:options.y1,
        x2:options.x2,
        y2:options.y2,
        x3:options.x3,
        y3:options.y3,
        color:options.color,
        type: options.type,
        draw(options?: TriangleOptions) { return draw(this, data, options) },
        setAngle1(x, y) {return setAngle(this, x, y, 1)},
        setAngle2(x, y) {return setAngle(this, x, y, 2)},
        setAngle3(x, y) {return setAngle(this, x, y, 3)},
        setType(type) { return setType(this, type) as TriangleReturn },
        setRGB(red, green, blue) { return setRGB(this, red, green, blue) as TriangleReturn },
        setRGBA(red, green, blue, alpha) { return setRGBA(this, red, green, blue, alpha) as TriangleReturn },
        setGradient(gradient) { return setGradient(this, gradient) as TriangleReturn },
        setColor(color) { return setColor(this, color) as TriangleReturn }
    }
}

function draw(triangle: TriangleReturn, data: Shape[], options?: TriangleOptions): TriangleReturn {
    if(!options) options ={};
    if(!options.x1) options.x1 = options.x1 == 0 ? 0 : options.x1 || triangle.x1;
    if(!options.y1) options.y1 = options.y1 == 0 ? 0 : options.y1 || triangle.y1;
    if(!options.x2) options.x2 = options.x2 == 0 ? 0 : options.x2 || triangle.x2;
    if(!options.y2) options.y2 = options.y2 == 0 ? 0 : options.y2 || triangle.y2;
    if(!options.x3) options.x3 = options.x3 == 0 ? 0 : options.x3 || triangle.x3;
    if(!options.y3) options.y3 = options.y3 == 0 ? 0 : options.y3 || triangle.y3;
    if(!options.color) options.color = options.color ? options.color : triangle.color;
    if(!options.type) options.type =  options.type ? options.type : triangle.type;
    data.push({
        shape_type: "Triangle",
        data: [options.x1, options.y1, options.x2, options.y2, options.x3, options.y3],
        draw_type: options.type == "fill" ? "Fill" : "Stroke",
        text: null,
        path: null,
        color: options.color
    });
    return triangle
}